File: shared/agent/improve_loop.py
"""
Autonomous Claude improve loop.

Runs overnight:
  1. Plays N local sim games → collects score distribution + replays
  2. Reads replays → asks Claude to identify failure patterns
  3. Claude proposes a targeted code fix
  4. Evaluates fix (N more games) → commits if score improved, reverts if not
  5. Repeats

Usage:
  python improve_loop.py \
    --bot-dir ../../challenges/groceries/bot \
    --sim-script ../../challenges/groceries/sim/run_sim.py \
    --games-per-eval 20 \
    --max-iterations 50 \
    --min-improvement 2
"""
from __future__ import annotations
import asyncio
import json
import subprocess
import shutil
import logging
import argparse
import time
from pathlib import Path
from datetime import datetime

import anthropic

logger = logging.getLogger("improve_loop")
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
)

# Model for code analysis and improvement
MODEL = "claude-opus-4-6"

client = anthropic.Anthropic()


def run_sim_games(sim_script: str, bot_dir: str, n_games: int, difficulty: str = "medium") -> list[int]:
    """
    Runs n_games in the local simulator and returns a list of scores.
    The simulator script should accept --bot-dir, --games, --difficulty flags
    and print one score per line to stdout.
    """
    result = subprocess.run(
        ["python", sim_script, "--bot-dir", bot_dir, "--games", str(n_games), "--difficulty", difficulty],
        capture_output=True,
        text=True,
        timeout=300,
    )
    if result.returncode != 0:
        logger.error(f"Sim failed: {result.stderr}")
        return []
    scores = []
    for line in result.stdout.strip().split("\n"):
        try:
            scores.append(int(line.strip()))
        except ValueError:
            pass
    return scores


def median_score(scores: list[int]) -> float:
    if not scores:
        return 0.0
    s = sorted(scores)
    n = len(s)
    if n % 2 == 0:
        return (s[n // 2 - 1] + s[n // 2]) / 2
    return float(s[n // 2])


def read_recent_replays(replay_dir: Path, n: int = 5) -> list[dict]:
    """Reads the N most recent replay files for analysis."""
    replays = sorted(replay_dir.glob("*.json"), key=lambda p: p.stat().st_mtime, reverse=True)
    result = []
    for path in replays[:n]:
        try:
            result.append(json.loads(path.read_text()))
        except Exception:
            pass
    return result


def read_bot_files(bot_dir: Path) -> dict[str, str]:
    """Reads all Python files in the bot directory."""
    files = {}
    for path in bot_dir.glob("*.py"):
        files[path.name] = path.read_text()
    return files


def analyze_replays_with_claude(replays: list[dict], bot_files: dict[str, str]) -> dict:
    """
    Asks Claude to analyze replays and identify the most impactful failure pattern.
    Returns {"pattern": str, "file": str, "fix": str, "code": str}
    """
    # Summarize replays (don't send entire replay — too large)
    replay_summaries = []
    for replay in replays:
        if not replay:
            continue
        rounds = replay
        total_rounds = len(rounds)
        final_score = rounds[-1].get("score", 0) if rounds else 0

        # Find rounds where bots waited when they shouldn't have
        wait_rounds = sum(
            1 for r in rounds
            if any(a.get("action") == "wait" for a in r.get("actions", []))
        )
        # Find max planning time
        max_planning_ms = max((r.get("planning_ms", 0) for r in rounds), default=0)

        replay_summaries.append({
            "total_rounds": total_rounds,
            "final_score": final_score,
            "wait_rounds": wait_rounds,
            "wait_pct": round(wait_rounds / max(total_rounds, 1) * 100, 1),
            "max_planning_ms": max_planning_ms,
            "sample_rounds": rounds[::max(1, total_rounds // 10)][:10],  # every 10th round
        })

    prompt = f"""You are analyzing an AI bot for a grocery delivery game competition.

The bot controls multiple robots that pick up items from shelves and deliver them to drop-off zones.
Scoring: +1 per item delivered, +5 per complete order. More completed orders = better score.

## Bot source code

{json.dumps({k: v for k, v in bot_files.items()}, indent=2)}

## Recent game replay summaries (last {len(replay_summaries)} games)

{json.dumps(replay_summaries, indent=2)}

## Your task

1. Identify the SINGLE most impactful failure pattern (the one that costs the most points)
2. Propose a SPECIFIC, MINIMAL code fix (edit ONE function if possible)
3. Explain exactly what to change and why it will help

Focus on real problems visible in the data:
- Bots waiting when they could be collecting or delivering
- Bots delivering with only 1 item when more are available
- Deadlocks in narrow aisles
- Poor task assignment (multiple bots chasing same item)
- Not using preview order to pre-fetch

Respond with a JSON object:
{{
  "failure_pattern": "one-sentence description of the main problem",
  "impact": "estimated points lost per game",
  "file": "which .py file to edit",
  "function": "which function to modify",
  "explanation": "why this fix helps",
  "new_code": "the complete replacement function (Python code as string)"
}}"""

    response = client.messages.create(
        model=MODEL,
        max_tokens=4096,
        thinking={"type": "adaptive"},
        messages=[{"role": "user", "content": prompt}],
    )

    text = next((b.text for b in response.content if b.type == "text"), "")

    # Extract JSON from response
    import re
    json_match = re.search(r'\{.*\}', text, re.DOTALL)
    if json_match:
        try:
            return json.loads(json_match.group())
        except json.JSONDecodeError:
            pass

    logger.warning("Claude did not return valid JSON")
    return {}


def apply_fix(bot_dir: Path, fix: dict) -> bool:
    """
    Applies the proposed fix to the bot code.
    Returns True if applied successfully.
    """
    file_name = fix.get("file")
    new_code = fix.get("new_code")
    function_name = fix.get("function")

    if not file_name or not new_code or not function_name:
        logger.warning("Incomplete fix spec")
        return False

    target = bot_dir / file_name
    if not target.exists():
        logger.warning(f"Target file not found: {target}")
        return False

    original = target.read_text()

    # Create backup
    backup = target.with_suffix(".py.bak")
    backup.write_text(original)

    # Replace the function
    import ast
    try:
        # Try to find and replace the function in the source
        lines = original.split("\n")
        result_lines = _replace_function(lines, function_name, new_code)
        if result_lines:
            target.write_text("\n".join(result_lines))
            logger.info(f"Applied fix to {file_name}:{function_name}")
            return True
        else:
            logger.warning(f"Could not locate function {function_name} in {file_name}")
            return False
    except Exception as e:
        logger.error(f"Failed to apply fix: {e}")
        backup.replace(target)  # restore
        return False


def _replace_function(lines: list[str], func_name: str, new_code: str) -> list[str] | None:
    """
    Finds `def func_name` in lines and replaces the entire function with new_code.
    Returns modified lines or None if function not found.
    """
    # Find the start of the function
    start_idx = None
    for i, line in enumerate(lines):
        stripped = line.lstrip()
        if stripped.startswith(f"def {func_name}(") or stripped.startswith(f"async def {func_name}("):
            start_idx = i
            break

    if start_idx is None:
        return None

    # Find the end of the function (next def at same or lower indent level)
    base_indent = len(lines[start_idx]) - len(lines[start_idx].lstrip())
    end_idx = len(lines)
    for i in range(start_idx + 1, len(lines)):
        line = lines[i]
        if not line.strip():
            continue
        indent = len(line) - len(line.lstrip())
        if indent <= base_indent and (line.lstrip().startswith("def ") or
                                       line.lstrip().startswith("async def ") or
                                       line.lstrip().startswith("class ")):
            end_idx = i
            break

    new_lines = lines[:start_idx] + new_code.split("\n") + [""] + lines[end_idx:]
    return new_lines


def revert_fix(bot_dir: Path, file_name: str):
    """Restores the backup file."""
    target = bot_dir / file_name
    backup = target.with_suffix(".py.bak")
    if backup.exists():
        backup.replace(target)
        logger.info(f"Reverted {file_name}")


async def run_improve_loop(
    bot_dir: str,
    sim_script: str,
    replay_dir: str,
    games_per_eval: int = 20,
    max_iterations: int = 50,
    min_improvement: float = 2.0,
    difficulty: str = "medium",
):
    bot_path = Path(bot_dir)
    replay_path = Path(replay_dir)
    history = []

    logger.info(f"Starting improve loop: {max_iterations} iterations, {games_per_eval} games/eval")

    # Baseline
    logger.info("Running baseline evaluation...")
    baseline_scores = run_sim_games(sim_script, bot_dir, games_per_eval, difficulty)
    baseline = median_score(baseline_scores)
    logger.info(f"Baseline median score: {baseline:.1f} (n={len(baseline_scores)})")

    current_best = baseline

    for iteration in range(max_iterations):
        logger.info(f"\n{'='*60}")
        logger.info(f"Iteration {iteration + 1}/{max_iterations} | Best: {current_best:.1f}")

        # Read recent replays and bot code
        replays = read_recent_replays(replay_path, n=5)
        bot_files = read_bot_files(bot_path)

        # Ask Claude for a fix
        logger.info("Asking Claude for improvement...")
        fix = analyze_replays_with_claude(replays, bot_files)

        if not fix:
            logger.warning("No fix proposed — skipping iteration")
            continue

        logger.info(f"Failure pattern: {fix.get('failure_pattern', 'unknown')}")
        logger.info(f"Proposed fix: {fix.get('file')}:{fix.get('function')}")
        logger.info(f"Explanation: {fix.get('explanation', '')}")

        # Apply the fix
        applied = apply_fix(bot_path, fix)
        if not applied:
            continue

        # Evaluate the fix
        logger.info(f"Evaluating fix ({games_per_eval} games)...")
        new_scores = run_sim_games(sim_script, bot_dir, games_per_eval, difficulty)
        new_median = median_score(new_scores)
        improvement = new_median - current_best

        logger.info(f"New median: {new_median:.1f} | Improvement: {improvement:+.1f}")

        if improvement >= min_improvement:
            # Keep the fix
            current_best = new_median
            logger.info(f"✅ Fix accepted! New best: {current_best:.1f}")
            # Clean up backup
            backup = bot_path / (fix.get("file", "").replace(".py", ".py.bak"))
            if backup.exists():
                backup.unlink()
        else:
            # Revert
            revert_fix(bot_path, fix.get("file", ""))
            logger.info(f"❌ Fix rejected (improvement {improvement:+.1f} < threshold {min_improvement})")

        history.append({
            "iteration": iteration + 1,
            "timestamp": datetime.now().isoformat(),
            "pattern": fix.get("failure_pattern"),
            "fix": f"{fix.get('file')}:{fix.get('function')}",
            "before": current_best if improvement < min_improvement else current_best - improvement,
            "after": new_median,
            "accepted": improvement >= min_improvement,
        })

        # Save history
        history_file = bot_path.parent / "improve_history.json"
        history_file.write_text(json.dumps(history, indent=2))

    logger.info(f"\n{'='*60}")
    logger.info(f"Improve loop complete. Final best: {current_best:.1f} (started: {baseline:.1f})")
    logger.info(f"Total improvement: {current_best - baseline:+.1f} points")


def main():
    parser = argparse.ArgumentParser(description="NM i AI 2026 — Autonomous Improve Loop")
    parser.add_argument("--bot-dir", required=True)
    parser.add_argument("--sim-script", required=True)
    parser.add_argument("--replay-dir", default="../../challenges/groceries/replays")
    parser.add_argument("--games-per-eval", type=int, default=20)
    parser.add_argument("--max-iterations", type=int, default=50)
    parser.add_argument("--min-improvement", type=float, default=2.0)
    parser.add_argument("--difficulty", default="medium")
    args = parser.parse_args()

    asyncio.run(run_improve_loop(
        bot_dir=args.bot_dir,
        sim_script=args.sim_script,
        replay_dir=args.replay_dir,
        games_per_eval=args.games_per_eval,
        max_iterations=args.max_iterations,
        min_improvement=args.min_improvement,
        difficulty=args.difficulty,
    ))


if __name__ == "__main__":
    main()

File: shared/agent/improve-loop.ts
import { execFileSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import Anthropic from "@anthropic-ai/sdk";
import minimist from "minimist";

function log(level: string, message: string): void {
  const timestamp = new Date().toISOString();
  console.error(`${timestamp} [${level}] ${message}`);
}

const MODEL = "claude-opus-4-6";
const client = new Anthropic();

function runSimGames(
  simScript: string,
  botDir: string,
  nGames: number,
  difficulty = "medium",
): number[] {
  try {
    const result = execFileSync(
      "node",
      ["--experimental-transform-types", simScript, "--bot-dir", botDir, "--games", String(nGames), "--difficulty", difficulty],
      { timeout: 300_000, encoding: "utf-8" },
    );
    const scores: number[] = [];
    for (const line of result.trim().split("\n")) {
      const num = parseInt(line.trim(), 10);
      if (!isNaN(num)) scores.push(num);
    }
    return scores;
  } catch (err: any) {
    log("ERROR", `Sim failed: ${err.stderr ?? err.message}`);
    return [];
  }
}

function medianScore(scores: number[]): number {
  if (scores.length === 0) return 0;
  const sorted = [...scores].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  if (sorted.length % 2 === 0) {
    return (sorted[mid - 1] + sorted[mid]) / 2;
  }
  return sorted[mid];
}

function readRecentReplays(replayDir: string, count = 5): any[] {
  const dir = replayDir;
  if (!fs.existsSync(dir)) return [];

  const files = fs
    .readdirSync(dir)
    .filter((f) => f.endsWith(".json"))
    .map((f) => ({ name: f, mtime: fs.statSync(path.join(dir, f)).mtimeMs }))
    .sort((a, b) => b.mtime - a.mtime)
    .slice(0, count);

  const replays: any[] = [];
  for (const file of files) {
    try {
      replays.push(JSON.parse(fs.readFileSync(path.join(dir, file.name), "utf-8")));
    } catch {
      // skip corrupt files
    }
  }
  return replays;
}

function readBotFiles(botDir: string): Record<string, string> {
  const files: Record<string, string> = {};
  for (const entry of fs.readdirSync(botDir)) {
    if (entry.endsWith(".ts") && !entry.endsWith(".test.ts")) {
      files[entry] = fs.readFileSync(path.join(botDir, entry), "utf-8");
    }
  }
  return files;
}

async function analyzeReplaysWithClaude(
  replays: any[],
  botFiles: Record<string, string>,
): Promise<Record<string, any>> {
  const replaySummaries: any[] = [];
  for (const replay of replays) {
    if (!replay || !Array.isArray(replay)) continue;
    const rounds = replay;
    const totalRounds = rounds.length;
    const finalScore = rounds[totalRounds - 1]?.score ?? 0;

    const waitRounds = rounds.filter((r: any) =>
      r.actions?.some((a: any) => a.action === "wait"),
    ).length;

    const maxPlanningMs = Math.max(
      ...rounds.map((r: any) => r.planning_ms ?? 0),
      0,
    );

    const step = Math.max(1, Math.floor(totalRounds / 10));
    const sampleRounds = rounds.filter((_: any, idx: number) => idx % step === 0).slice(0, 10);

    replaySummaries.push({
      total_rounds: totalRounds,
      final_score: finalScore,
      wait_rounds: waitRounds,
      wait_pct: Math.round((waitRounds / Math.max(totalRounds, 1)) * 1000) / 10,
      max_planning_ms: maxPlanningMs,
      sample_rounds: sampleRounds,
    });
  }

  const prompt = `You are analyzing an AI bot for a grocery delivery game competition.

The bot controls multiple robots that pick up items from shelves and deliver them to drop-off zones.
Scoring: +1 per item delivered, +5 per complete order. More completed orders = better score.

## Bot source code

${JSON.stringify(botFiles, null, 2)}

## Recent game replay summaries (last ${replaySummaries.length} games)

${JSON.stringify(replaySummaries, null, 2)}

## Your task

1. Identify the SINGLE most impactful failure pattern (the one that costs the most points)
2. Propose a SPECIFIC, MINIMAL code fix (edit ONE function if possible)
3. Explain exactly what to change and why it will help

Focus on real problems visible in the data:
- Bots waiting when they could be collecting or delivering
- Bots delivering with only 1 item when more are available
- Deadlocks in narrow aisles
- Poor task assignment (multiple bots chasing same item)
- Not using preview order to pre-fetch

Respond with a JSON object:
{
  "failure_pattern": "one-sentence description of the main problem",
  "impact": "estimated points lost per game",
  "file": "which .ts file to edit",
  "function": "which function to modify",
  "explanation": "why this fix helps",
  "new_code": "the complete replacement function (TypeScript code as string)"
}`;

  const response = await client.messages.create({
    model: MODEL,
    max_tokens: 4096,
    thinking: { type: "enabled", budget_tokens: 2048 },
    messages: [{ role: "user", content: prompt }],
  });

  const text = response.content.find((b) => b.type === "text")?.text ?? "";

  const jsonMatch = text.match(/\{[\s\S]*\}/);
  if (jsonMatch) {
    try {
      return JSON.parse(jsonMatch[0]);
    } catch {
      // fall through
    }
  }

  log("WARN", "Claude did not return valid JSON");
  return {};
}

function replaceFunction(
  lines: string[],
  funcName: string,
  newCode: string,
): string[] | null {
  let startIdx: number | null = null;
  for (let idx = 0; idx < lines.length; idx++) {
    const stripped = lines[idx].trimStart();
    if (
      stripped.startsWith(`function ${funcName}(`) ||
      stripped.startsWith(`export function ${funcName}(`) ||
      stripped.startsWith(`async function ${funcName}(`) ||
      stripped.startsWith(`export async function ${funcName}(`)
    ) {
      startIdx = idx;
      break;
    }
  }

  if (startIdx === null) return null;

  const baseIndent = lines[startIdx].length - lines[startIdx].trimStart().length;
  let endIdx = lines.length;
  for (let idx = startIdx + 1; idx < lines.length; idx++) {
    const line = lines[idx];
    if (!line.trim()) continue;
    const indent = line.length - line.trimStart().length;
    if (
      indent <= baseIndent &&
      (line.trimStart().startsWith("function ") ||
        line.trimStart().startsWith("export function ") ||
        line.trimStart().startsWith("async function ") ||
        line.trimStart().startsWith("export async function ") ||
        line.trimStart().startsWith("export class ") ||
        line.trimStart().startsWith("class "))
    ) {
      endIdx = idx;
      break;
    }
  }

  return [...lines.slice(0, startIdx), ...newCode.split("\n"), "", ...lines.slice(endIdx)];
}

function applyFix(botDir: string, fix: Record<string, any>): boolean {
  const fileName = fix.file;
  const newCode = fix.new_code;
  const functionName = fix.function;

  if (!fileName || !newCode || !functionName) {
    log("WARN", "Incomplete fix spec");
    return false;
  }

  const target = path.join(botDir, fileName);
  if (!fs.existsSync(target)) {
    log("WARN", `Target file not found: ${target}`);
    return false;
  }

  const original = fs.readFileSync(target, "utf-8");
  const backup = target.replace(/\.ts$/, ".ts.bak");
  fs.writeFileSync(backup, original);

  try {
    const lines = original.split("\n");
    const resultLines = replaceFunction(lines, functionName, newCode);
    if (resultLines) {
      fs.writeFileSync(target, resultLines.join("\n"));
      log("INFO", `Applied fix to ${fileName}:${functionName}`);
      return true;
    }
    log("WARN", `Could not locate function ${functionName} in ${fileName}`);
    return false;
  } catch (err: any) {
    log("ERROR", `Failed to apply fix: ${err.message}`);
    // Restore backup
    fs.copyFileSync(backup, target);
    return false;
  }
}

function revertFix(botDir: string, fileName: string): void {
  const target = path.join(botDir, fileName);
  const backup = target.replace(/\.ts$/, ".ts.bak");
  if (fs.existsSync(backup)) {
    fs.copyFileSync(backup, target);
    fs.unlinkSync(backup);
    log("INFO", `Reverted ${fileName}`);
  }
}

async function runImproveLoop(opts: {
  botDir: string;
  simScript: string;
  replayDir: string;
  gamesPerEval: number;
  maxIterations: number;
  minImprovement: number;
  difficulty: string;
}): Promise<void> {
  const history: any[] = [];

  log("INFO", `Starting improve loop: ${opts.maxIterations} iterations, ${opts.gamesPerEval} games/eval`);

  // Baseline
  log("INFO", "Running baseline evaluation...");
  const baselineScores = runSimGames(opts.simScript, opts.botDir, opts.gamesPerEval, opts.difficulty);
  const baseline = medianScore(baselineScores);
  log("INFO", `Baseline median score: ${baseline.toFixed(1)} (n=${baselineScores.length})`);

  let currentBest = baseline;

  for (let iteration = 0; iteration < opts.maxIterations; iteration++) {
    log("INFO", `\n${"=".repeat(60)}`);
    log("INFO", `Iteration ${iteration + 1}/${opts.maxIterations} | Best: ${currentBest.toFixed(1)}`);

    const replays = readRecentReplays(opts.replayDir, 5);
    const botFiles = readBotFiles(opts.botDir);

    log("INFO", "Asking Claude for improvement...");
    const fix = await analyzeReplaysWithClaude(replays, botFiles);

    if (!fix || Object.keys(fix).length === 0) {
      log("WARN", "No fix proposed — skipping iteration");
      continue;
    }

    log("INFO", `Failure pattern: ${fix.failure_pattern ?? "unknown"}`);
    log("INFO", `Proposed fix: ${fix.file}:${fix.function}`);
    log("INFO", `Explanation: ${fix.explanation ?? ""}`);

    const applied = applyFix(opts.botDir, fix);
    if (!applied) continue;

    log("INFO", `Evaluating fix (${opts.gamesPerEval} games)...`);
    const newScores = runSimGames(opts.simScript, opts.botDir, opts.gamesPerEval, opts.difficulty);
    const newMedian = medianScore(newScores);
    const improvement = newMedian - currentBest;

    log("INFO", `New median: ${newMedian.toFixed(1)} | Improvement: ${improvement >= 0 ? "+" : ""}${improvement.toFixed(1)}`);

    if (improvement >= opts.minImprovement) {
      currentBest = newMedian;
      log("INFO", `Fix accepted! New best: ${currentBest.toFixed(1)}`);
      const backup = path.join(opts.botDir, (fix.file ?? "").replace(/\.ts$/, ".ts.bak"));
      if (fs.existsSync(backup)) fs.unlinkSync(backup);
    } else {
      revertFix(opts.botDir, fix.file ?? "");
      log("INFO", `Fix rejected (improvement ${improvement >= 0 ? "+" : ""}${improvement.toFixed(1)} < threshold ${opts.minImprovement})`);
    }

    history.push({
      iteration: iteration + 1,
      timestamp: new Date().toISOString(),
      pattern: fix.failure_pattern,
      fix: `${fix.file}:${fix.function}`,
      before: improvement >= opts.minImprovement ? currentBest - improvement : currentBest,
      after: newMedian,
      accepted: improvement >= opts.minImprovement,
    });

    const historyFile = path.join(path.dirname(opts.botDir), "improve_history.json");
    fs.writeFileSync(historyFile, JSON.stringify(history, null, 2));
  }

  log("INFO", `\n${"=".repeat(60)}`);
  log("INFO", `Improve loop complete. Final best: ${currentBest.toFixed(1)} (started: ${baseline.toFixed(1)})`);
  log("INFO", `Total improvement: ${currentBest - baseline >= 0 ? "+" : ""}${(currentBest - baseline).toFixed(1)} points`);
}

function main(): void {
  const args = minimist(process.argv.slice(2), {
    string: ["bot-dir", "sim-script", "replay-dir", "difficulty"],
    default: {
      "replay-dir": "../../challenges/groceries/replays",
      "games-per-eval": 20,
      "max-iterations": 50,
      "min-improvement": 2,
      difficulty: "medium",
    },
  });

  if (!args["bot-dir"] || !args["sim-script"]) {
    console.error(
      "Usage: node --experimental-transform-types improve-loop.ts " +
        "--bot-dir DIR --sim-script SCRIPT [--games-per-eval 20] " +
        "[--max-iterations 50] [--min-improvement 2] [--difficulty medium]",
    );
    process.exit(1);
  }

  runImproveLoop({
    botDir: args["bot-dir"],
    simScript: args["sim-script"],
    replayDir: args["replay-dir"],
    gamesPerEval: args["games-per-eval"],
    maxIterations: args["max-iterations"],
    minImprovement: args["min-improvement"],
    difficulty: args.difficulty,
  }).catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

main();
